Feature 914 support sierra experience events

sawhneyh opened the pull request:
PR raised for Issue: https://github.com/adobe/xdm/issues/914

sawhneyh commented:
I have run the lint locally, and there were whitesapces issues in the existing files. The build is failing for the lint error on a file, which is not part of this PR.

sawhneyh commented on extensions/adobe/experience/adcloud/searchads/account.example.1.json:
```diff
@@ -6,7 +6,7 @@
   "searchads:budget": 123.45,
   "searchads:budgetType": "Monthly",
-  "searchads:searchEngineId: 3,
+  "searchads:searchEngineId": 3,
```
Existing lint issue, prior to this PR.

sawhneyh commented on extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json:
```diff
@@ -0,0 +1,58 @@
+    "you may not use this file except in compliance with the License. You may obtain a copy",
+    "of the License at https://creativecommons.org/licenses/by/4.0/"
+  ],
+  "$id": "https://ns.adobe.com/xdm/experience/customerJourneyManagement/email-channel-context",
```
All the schema extensions are created under the hierarchy "adobe/experience/customerJourneyManagement" but the "id" field has "xdm" in path. This was require, as per the test scripts the id field has to follow the regex:
`should match pattern "(https://ns\.adobe\.com/xdm/[a-z0-9-/]*)|(http://schema\.org/.*)|(http://ns.adobe.com/adobecloud/core/1.0.*)|([http|https]://ns\.adobe\.com/adobecloudplatform/.*)|(https://tools\.ietf\.org/html/draft-kelly-json-hal-08/.*)|(http://www.iptc\.org/.*)|(https://id3\.org/id3v2.4/.*)`

sawhneyh commented:
The current test scripts, doesn't verify the schemas and examples under the "/extension" folder. For testing the schema i have to modify the following scripts to validate the schemas and example added with this script. We should have separate issue for fixing the following scripts.

1. test/example-valid.js
2. test/example-presence.js
3. test/invalid-presence.js

sawhneyh commented on extensions/adobe/experience/customerJourneyManagement/messageexecution.schema.json:
```diff
@@ -0,0 +1,58 @@
+          "type": "string",
+          "description": "The ID of the external Application entity triggering the MessageExecution."
+        },
+        "xdm:externalApplication": {
```
Should we have dedicated field like JourneyId for referring to the external application which initiated the message execution. Or keep it generic with field 'xdm:externalApplicationID'.

jfait replied:
Could the ID be namespaced?
IIRC XDM had a datasource or similar that had notion of NS+ID

ctrubian replied:
the combination of externalAppicationID and externalApplication provides a namespaced ID

ctrubian replied:
and I prefer the generic approach rather than a journey specific one

jfait commented on extensions/adobe/experience/customerJourneyManagement/message-delivery-feedback.schema.json:
```diff
@@ -0,0 +1,96 @@
+              "type": "string",
+              "description": "Classifies whether it is a soft bounce or hard bounce.",
+              "meta:enum": {
+                "soft": "The reason for failed delivery is temprorary.",
```
temporary

jfait commented on extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.1.json:
```diff
@@ -0,0 +1,40 @@
+  "xdm:messageProfile": {
+    "@id": "4218b775-bef3-46b2-aee2-7caae052cf94",
+    "xdm:channel": {
+      "@id": "https://ns.adobe.com/xdm/channels/email"
```
do we need this given xdm:channelConext further below

sawhneyh replied:
In xdm:channelConext we don't have any channelType field to specify for which channel is this EE for.  The consumer of this schema has to rely on presence of object like "xdm:emailChannelContext" or "xdm:pushChannelContext" to identify whether this messages is for Email or Push channel.

Though we can add an explicit channel type field in "xdm:channelConext", but for completeness of messageProfile mixin which can be used in other schemas(if required) also, I added this property here.

jfait commented on extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.4.json:
```diff
@@ -0,0 +1,44 @@
+{
```
should we name the examples to indicate;
EVENT_TYPE & CHANNEL & STATUS

this one looks like feedback-push-exclude example for instance

sawhneyh replied:
The current automation scripts which validates the examples with the corresponding schema file requires, example files to have a naming convention : schema-name.example.[0-9].json. Check line:
1. https://github.com/adobe/xdm/blob/master/test/example-valid.js#L13
2. https://github.com/adobe/xdm/blob/master/test/example-valid.js#L67

jfait commented on extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.5.json:
```diff
@@ -0,0 +1,45 @@
+{
```
should we name the examples to indicate;
EVENT_TYPE & CHANNEL & STATUS

this one looks like tracking-email-click example for instance

sawhneyh replied:
The current automation scripts which validates the examples with the corresponding schema file requires, example files to have a naming convention : schema-name.example.[0-9].json

jfait commented on extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.5.json:
```diff
@@ -0,0 +1,45 @@
+  "xdm:messageExecution": {
+    "@id": "4218b775-bef3-46b2-aee2-7caae052cf94",
+    "xdm:messageID": "aa440b96-ce65-4ad2-b95d-8c3b51f89bc2",
+    "xdm:messageLabel": "Deal for Gold members",
```
Do need message label?
Can labels on a message change overtime? any impact?

sawhneyh replied:
I added this for the use cases where customer will be building the segmentation/reporting use cases in AEP/CJA. For usability purpose i added the label fields, otherwise it will be difficult for end user to build segments/reports using the machine generated ids.

jfait commented on extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.6.json:
```diff
@@ -0,0 +1,39 @@
+  },
+  "xdm:eventStatus": {
+    "xdm:messageInteraction": {
+      "xdm:interactionType": "action"
```
Are there action details missing here?

sawhneyh replied:
There are no specific requirement provided for the reporting or segmentation use cases on the metadata required with the "action" event.
Though will still check with Pdms for an explicit requirements.
cc: @aashishjassal

jfait commented on extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.6.json:
```diff
@@ -0,0 +1,39 @@
+  "xdm:messageProfile": {
+    "@id": "4218b775-bef3-46b2-aee2-7caae052cf94",
+    "xdm:channel": {
+      "@id": "https://ns.adobe.com/xdm/channels/email"
```
this looks like a push event but channel is email

jfait commented on extensions/adobe/experience/customerJourneyManagement/message-interaction.schema.json:
```diff
@@ -0,0 +1,60 @@
+          "type": "string",
+          "description": "Unique Identifier of the Link clicked by the user."
+        },
+        "xdm:label": {
```
Do we need label? is this for user friendly name display in reporting or is there any other reason to carry forward labels?

ctrubian replied:
+1 what is this label and where does it come from?

sawhneyh replied:
> is this for user friendly name display in reporting or is there any other reason to carry forward labels?

Yes this is for reporting use cases in CJA.

> what is this label and where does it come from?

Enrichment service will enrich this using the link id.

ctrubian replied:
So this is something that will be set in the authoring environment and that will propagate it to somewhere that enrichment can find it?

jfait commented on extensions/adobe/experience/customerJourneyManagement/messageprofile.schema.json:
```diff
@@ -0,0 +1,48 @@
+          "$ref": "https://ns.adobe.com/xdm/channels/channel"
+        },
+        "xdm:variant": {
+          "title": "Content Variant",
```
Can variant be more complex and nested?

What if message header had 4 variants, message body had 2, and message footer had 3.

Is that possible or would variants be a cumulative sum of outcomes?

sawhneyh replied:
I am not aware of this requirement, if we have this requirement then we should add a support for defining multiple variants with respective placement.

aashishjassal replied:
As far as I also know there is a variant at a message profile level, complex nesting is there. @jfait Are you aware of any such possible requirements?

jfait commented on extensions/adobe/experience/customerJourneyManagement/push-channel-context.example.1.json:
```diff
@@ -0,0 +1,5 @@
+{
```
Mobile team should also review if they haven't already

aashishjassal replied:
There is an ambiguity on Mobile team right now. We are waiting for some discussion to finalize and will then float it to them

ctrubian commented on extensions/adobe/experience/customerJourneyManagement/message-interaction.schema.json:
```diff
@@ -0,0 +1,60 @@
+          }
+        },
+        "@id": {
+          "title": "Tracking URL ID",
```
Nit picky, but can we change this to "Tracker link ID"

ctrubian commented on extensions/adobe/experience/customerJourneyManagement/message-interaction.schema.json:
```diff
@@ -0,0 +1,60 @@
+          "description": "The user-friendly label clicking on which the link opens."
+        },
+        "xdm:category": {
+          "title": "URL Category",
```
"Link Category"

ctrubian commented:
In general this looks good to me for tracker purposes.  Just spelling/wording and removal or explanation of what the link label is.

cmathis commented on extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json:
```diff
@@ -0,0 +1,58 @@
+  "definitions": {
+    "emailChannelContext": {
+      "properties": {
+        "xdm:address": {
```
Because this is not a core standard XDM mixin, these fields should not be in the XDM namespace either. This field should be named "https://ns.adobe.com/xdm/experience/customerJourneyManagement/address"

sawhneyh replied:
Hi @cmathis:  If we rename the field "https://ns.adobe.com/xdm/experience/customerJourneyManagement/address", then while ingesting or accessing the data, do we need to use the complete path, or it get stripped of while accessing the data through xed format?

cc: @aashishjassal

sawhneyh replied:
@cmathis : Can you explain when we have to use the xdm namespace and when to use the complete uri path for defining the properties in mixin.

There are existing mixin like https://github.com/adobe/xdm/blob/master/extensions/adobe/experience/campaign/feedbackevent.schema.json which have properties defined using:

1. xdm namespance
2. uri style namespace
3. No namespaces.

cmathis replied:
Every field in XDM has a namespace. The URI style is the long form when you do not have a registered shorthand such as xdm, repo, adcloud, etc. There is no registered "customerJourneyManagement" short name so you would use the longer URI form which is fine as well. 

The path to this address field in AEP would become "_experience/customerJourneyManagmenet/address"

sawhneyh replied:
@cmathis : So if we are using the uri style namespaces for every field, and have a case like "domain" is a sub property of a "address" field. Then how will the path be mapped when we fetch the xdm given both "domain" and "address" have the same namespace path.
Do we have to access as: _experience/customerJourneyManagmenet/address. _experience/customerJourneyManagmenet/domain, or it will be _experience/customerJourneyManagmenet/address.domain as domain is a sub property of a address with same prefix namespace.

cmathis replied:
No the XED converter in AEP will recognize that the parent namespace is the same as the child namespace and will note require the duplication. It will be the second example - _experience/customerJourneyManagmenet/address.domain

cmathis commented on extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json:
```diff
@@ -0,0 +1,58 @@
+          "format": "email",
+          "description": "The technical address, for example, 'name@domain.com' as commonly defined in RFC2822 and subsequent standards."
+        },
+        "xdm:operationalContext": {
```
Field name should be "https://ns.adobe.com/xdm/experience/customerJourneyManagement/operationalContext"

cmathis commented on extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json:
```diff
@@ -0,0 +1,58 @@
+    "you may not use this file except in compliance with the License. You may obtain a copy",
+    "of the License at https://creativecommons.org/licenses/by/4.0/"
+  ],
+  "$id": "https://ns.adobe.com/xdm/experience/customerJourneyManagement/email-channel-context",
```
Remove the "xdm" from the $id path. Should be https://ns.adobe.com/experience/customerJourneyManagement/email-channel-context

sawhneyh replied:
@cmathis : I noticed current test scripts do not validate the schemas under the "extension" folder. For validating the schemas with examples in this PR, I have to modify the test scripts locally to validate them. 
While running the schemas validating scripts it gives the error that "id" should match the following regex, and because of that I added the "xdm" in the uri path.  I will remove the "xdm" from the uri.

`(https://ns\.adobe\.com/xdm/[a-z0-9-/]*)|(http://schema\.org/.*)|(http://ns.adobe.com/adobecloud/core/1.0.*)|([http|https]://ns\.adobe\.com/adobecloudplatform/.*)|(https://tools\.ietf\.org/html/draft-kelly-json-hal-08/.*)|(http://www.iptc\.org/.*)|(https://id3\.org/id3v2.4/.*)`


On a side note, I assume this is a known issue, that current validation scripts do not validate schemas under the extension folder.

cmathis commented:
For all new mixins and fields use the "https://ns.adobe.com/xdm/experience/customerJourneyManagement/" namespace.

sawhneyh commented on extensions/adobe/experience/customerJourneyManagement/message-experienceevent.example.1.json:
```diff
@@ -0,0 +1,45 @@
+  },
+  "xdm:directMarketing": {
+    "xdm:sends": {
+      "xdm:value": 1
```
Added explicit measure field for building the reports in CJA. In absence of this explicit fields we have to use the calculated metrics for the send/bounce/click/open metrics.

aashishjassal replied:
+1

sawhneyh commented:
@cmathis : You can continue the code review, I have incorporated the feedback to change the filed names to follow the uri pattern namespaces.

cmathis commented on extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json:
```diff
@@ -0,0 +1,110 @@
+        "https://ns.adobe.com/xdm/experience/customerJourneyManagement/channelContext": {
+          "title": "Message Channel context",
+          "description": "Message Channel fields for the Messageprofle associated with the Adobe CJM Experience Event.",
+          "oneOf": [
```
oneOf is not usable in AEP - you cannot have a field be two different data types.

sawhneyh replied:
@cmathis : removed the one of construct from here, but still using under the "required" properties to apply a constraint that "channelContext" can have either "emailChannelContext" or "pushChannelContext"

cmathis commented on extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json:
```diff
@@ -0,0 +1,110 @@
+      "$ref": "https://ns.adobe.com/xdm/common/extensible#/definitions/@context"
+    },
+    {
+      "$ref": "https://ns.adobe.com/xdm/context/experienceevent"
```
Mixins do not include the class in their allOf definition. The class will be included in the /schemas that is created along with this mixin.

cmathis replied:
So you want this to be a "global" cross tenant schema. This is technically possible, but not 100% supported in AEP. A user will not see this schema in the UI for example. Also not sure what will happen if they click on the schema of the Dataset. 

You can certainly try it though. To do that remove the "meta:intendedToExtend" (this is what makes it a mixin) and set the "meta:abstract" attribute to false.

cmathis commented on extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json:
```diff
@@ -0,0 +1,110 @@
+      "$ref": "https://ns.adobe.com/xdm/context/experienceevent"
+    },
+    {
+      "$ref": "https://ns.adobe.com/xdm/context/experienceevent-environment-details"
```
It appears you are including other mixins with this current definition. This is okay, but wanted to make sure that is what you wanted.

sawhneyh replied:
@cmathis : We plan to have "message-experienceevent.schema" as a schema rather a mixin. Is it possible to define and commit pre built schemas like "message-experienceevent.schema".

Also similarly we have used mixin within mixin at line: https://github.com/adobe/xdm/blob/56573ad19fa756030ade3494552c88f3c01543a6/extensions/adobe/experience/customerJourneyManagement/messageprofile.schema.json#L25. I assume, we support adding mixin within mixin.

cmathis replied:
You can have a mixin which reference other mixins so this is fine.

sawhneyh replied:
@cmathis : Also can you confirm first part of the query:

>  We plan to have "message-experienceevent.schema" as a schema rather a mixin. Is it possible to define and commit pre built schemas like "message-experienceevent.schema".

cmathis replied:
Yes, this is technically possible, but please schedule a meeting with XDM PM and myself to go over the details first.

prabhum2 commented:
@sawhneyh - It appears that there is more internal discussion going on this . Please let me know when you are ready for this PR is ready to get into master

cmathis commented on extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json:
```diff
@@ -0,0 +1,58 @@
+  "definitions": {
+    "emailChannelContext": {
+      "properties": {
+        "https://ns.adobe.com/xdm/experience/customerJourneyManagement/address": {
```
Need to remove the "xdm" from your URI string. Namespace should be https://ns.adobe.com/experience/customerJourneyManagement/. This is my fault because I had XDM in the example I showed in another comment.

prabhum2 commented:
@sawhneyh - Can you please resolve above conflict

sawhneyh commented:
@prabhum2 , @cmathis : You can resume the review, I have resolved the merged conflicts, and incorporated the review comments.

cmathis commented on extensions/adobe/experience/customerJourneyManagement/message-experienceevent.schema.json:
```diff
@@ -0,0 +1,114 @@
+              "description": "Push channel fields for the MessageProfile associated with the Adobe CJM Experience Event."
+            }
+          },
+           "oneOf": [
```
I'm checking, but pretty sure the AEP ingestion process will not support this. "oneOf" is a json-schema concept not really supported in AEP at this time.

cmathis replied:
I checked and have confirmed that this oneOf around the "required" attribute is not supported by the ingestion process. There is no way today to say you are required to have field X OR Y. If this is a problem, then please open a ticket with the data ingestion team.

sawhneyh replied:
Thanks @cmathis for update, I have removed the use of "oneOf" construct for defining the required properties.

sawhneyh replied:
@cmathis : You can resume the review. Also please provide your inputs regarding comment https://github.com/adobe/xdm/pull/915#discussion_r469872849.

We wanted to commit "message-experienceevent.schema" as a schema(not mixin which needs to be added over EE class) so that it can be available as OOTB schema in AEP.

If we have to make any changes for that, we will incorporate that.

cmathis commented:
@sawhneyh  - Let us know if there is still internal conversations going on with this PR. Other then the build failure, I think this PR looks good. Just waiting for your approval that it is ready to be officially merged in.

sawhneyh commented on extensions/adobe/experience/customerJourneyManagement/messageexecution.schema.json:
```diff
@@ -0,0 +1,53 @@
+  "definitions": {
+    "messageExecution": {
+      "properties": {
+        "@id": {
```
@cmathis : As per the https://github.com/adobe/xdm/blob/master/CONTRIBUTING.md we should not be inventing the new "id" attributes, but to use the "@id" convention.

Can you help usunderstand, how this "@id" field is namespaced, we may have a case where we add more than one mixin to schema, where both the mixjns has "@id" field. Then how this two "@id" fields are distinguished.

cc: @aashishjassal

cmathis replied:
The ExperienceEvent already includes an @id field, so this is basically ignored here. If you need an execution Id or something like it, then create a new field here named "executionId".

sawhneyh replied:
I have replaced the "@id" field  with the full name spaced field name.

cmathis commented on extensions/adobe/experience/customerJourneyManagement/email-channel-context.schema.json:
```diff
@@ -0,0 +1,60 @@
+              "title": "Outbound IP Address",
+              "type": "string",
+              "description": "Outbound IP address of the customer used to deliver the message.",
+              "oneOf": [
```
oneOf not supported in AEP. This will throw an error when you try to use the schema there.

sawhneyh commented:
1. I have fixed the lint issue.
2. Replaced the "@id" field with the proper name spaced field name.
3. Merged the email-channel-context and push-channel-context in messageProfile mixin.

@aashishjassal : Please review the changes.

aashishjassal commented:
Its approved, Please merge

sawhneyh commented:
@cmathis : can you please merge this PR. Also once we merge this PR, will these mixins automatically starts reflecting in AEP, or we need to raise some request for that.
Also will mixins be first deployed on AEP int environment or on all environment at once.

cmathis commented:
@sawhneyh  - These changes are currently installed on INT. They will be pushed to Stage on Monday and stay there for one week before they go to Prod. Please test things on INT and Stage next week.
